### PR TITLE
action: download default release assets to sign

### LINKED ISSUE #44: Figure out how to sign for the default release assets
By default, GitHub Actions produces `{tag}.tar.gz` and `{tag}.zip` assets for each release.

These artifacts are generated on the fly, meaning that they can (and have) changed their contents (and thus digests) over time. This recently caused some significant breakage due to incorrect assumptions around that: <https://github.blog/changelog/2023-01-30-git-archive-checksums-may-change/>

Given that they can change, the correct thing for us to do here is probably to download the asset as it appears, sign for it, and then re-upload it to the release as a new asset: this will freeze it, meaning that the signature will remain correct.

### PR DESCRIPTION
**Pull request by tnytown**
#### Summary

This PR resolves #44. We now do the following on releases if `release-signing-artifacts` is enabled:

1. Download the default release artifacts, i.e.`{{ref}}.tar.gz` and `{{ref}}.zip`;
2. Sign for them;
3. Upload them back to the release as `{{ref}}-signed.{{extension}}`, along with their signatures.

#### Release Note

The `release-signing-artifacts` setting was extended to re-upload and sign default assets. Previously, the stability of default asset hashes were not guaranteed, as they are generated on-the-fly by GitHub.

#### Documentation

https://docs.sigstore.dev does not seem to include documentation on the actions, but I will add a commit documenting the change to `release-signing-artifacts` in this repo's README.

### CODE REVIEW

**Comment by tnytown** [comment]
Doing this also significantly increases the number of assets attached to a release by default: https://github.com/tnytown/gh-action-sigstore-python/releases/tag/TEST_RELEASE

**Comment by woodruffw** [comment]
> 3\. Upload them back to the release as `{{ref}}-signed.{{extension}}`, along with their signatures.

What happens if we name it `{{ref}}.{{extension}}`, like the default artifacts? I'd partially expect that to fail, but if it works it might be slightly nicer from a consumer perspective (no need to understand why archives have "signed" in the name).

> https://docs.sigstore.dev/ does not seem to include documentation on the actions, but I will add a commit documenting the change to release-signing-artifacts in this repo's README.

Yep, the README is good -- this action isn't documented at all on the main Sigstore website. That's something we should consider in the future, though!

> Doing this also significantly increases the number of assets attached to a release by default: https://github.com/tnytown/gh-action-sigstore-python/releases/tag/TEST_RELEASE

Yeah -- the `.crt` and `.sig` outputs are technically redundant now that `.sigstore` is present, so I think we can add an option like `bundle-only: bool` with a default of `false` that'll control whether we emit those.

Then, with a 2.0 release, we can switch that setting to `true`.

File: action.py [hunk 1]
```diff
@@ -53,6 +53,34 @@ def _log(msg):
     print(msg, file=sys.stderr)
 
 
+def _download_ref_asset(ext):
+    repo = os.getenv('GITHUB_REPOSITORY')
+    ref = os.getenv("GITHUB_REF")
+
+    artifact = f"{os.getenv('GITHUB_REF_NAME')}-signed{ext}"
```
Comment (woodruffw) [inline action.py:60]: Nitpick: This shouldn't cause issues as-is, but let's use a `NamedTemporaryFile` or similar for this.

Comment (woodruffw) [reply action.py:60]: That, or just putting it under `/tmp` is probably sufficient 🙂

File: action.py [hunk 2]
```diff
@@ -53,6 +53,34 @@ def _log(msg):
...
+         "-f",
+         "-L",
+         "-o", str(artifact),
+         f"https://github.com/{repo}/archive/{ref}{ext}"],
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        # do not pass environment to curl
+    )
+    _debug(curl_status.stdout)
```
Comment (woodruffw) [inline action.py:75]: LGTM for the first blush, but in the interest of "native" requests let's use `requests`. 

Something roughly like:

```python
resp = requests.get("https://...")
resp.raise_for_status()

with Path(artifact).open("wb") as io:
    io.write(resp.content)
```

(You'll need to add `requests` to the `requirements.txt` as well, most likely.)

**Review by woodruffw — COMMENTED** [review]
LGTM overall! Two small nitpicks.

**Comment by woodruffw** [comment]
> Yeah -- the .crt and .sig outputs are technically redundant now that .sigstore is present, so I think we can add an option like bundle-only: bool with a default of false that'll control whether we emit those.

> Then, with a 2.0 release, we can switch that setting to true.

Just to clarify: We can do this with a follow-up PR. I'll make an issue to track that.

File: action.py (hunk 1 above)
Comment (tnytown) [reply action.py:60]: good point! will resolve this when I swap in `requests`

**Comment by tnytown** [comment]
> What happens if we name it `{{ref}}.{{extension}}`, like the default artifacts? I'd partially expect that to fail, but if it works it might be slightly nicer from a consumer perspective (no need to understand why archives have "signed" in the name).

I swapped it to `{{ref}}-signed.{{extension}}` when trying to fix another issue with artifact upload. I'll try `{{ref}}.{{extension}}` again to see if it works!

**Comment by woodruffw** [comment]
Cool! If it doesn't, we could always do something generic like `release.{{extension}}` -- the stable URL contains the `{{ref}}` anyways 🙂

File: action.py [hunk 3]
```diff
@@ -53,6 +55,22 @@ def _log(msg):
     print(msg, file=sys.stderr)
 
 
+def _download_ref_asset(ext):
+    repo = os.getenv('GITHUB_REPOSITORY')
+    ref = os.getenv("GITHUB_REF")
+
+    artifact = f"/tmp/{os.getenv('GITHUB_REF_NAME')}{ext}"
```
Comment (woodruffw) [inline action.py:62]:
```suggestion
    artifact = Path(f"/tmp/{os.getenv('GITHUB_REF_NAME')}").with_suffix(ext)
```

File: action.py [hunk 4]
```diff
@@ -53,6 +55,22 @@ def _log(msg):
...
+
+    # GitHub supports /:org/:repo/archive/:ref<.tar.gz|.zip>.
+    r = requests.get(f"https://github.com/{repo}/archive/{ref}{ext}", stream=True)
+    r.raise_for_status()
+    with Path(artifact).open("wb") as io:
```
Comment (woodruffw) [inline action.py:67]:
```suggestion
    with artifact.open("wb") as io:
```

File: action.py [hunk 5]
```diff
@@ -53,6 +55,22 @@ def _log(msg):
...
+        for chunk in r.iter_content(chunk_size=None):
+            io.write(chunk)
+
+    return artifact
```
Comment (woodruffw) [inline action.py:71]:
```suggestion
    return str(artifact)
```

**Review by woodruffw — DISMISSED** [review]
LGTM! Small tweaks but I'll merge those in and get this merged.

**Comment by woodruffw** [comment]
CI is broken because of third-party PRs not having access to OIDC, sigh. I'll tweak `all-tests-pass` so that it allows skippage.

**Review by woodruffw — APPROVED** [review]

**Pull request merged**